### DISCUSSION TITLE
Children of given Node of the Tree

### DISCUSSION THREAD

**Discussion by boblukas** [boostorg/geometry#1110] [Q&A]
Hi everyone,

for an application I need to know the hierarchy of my rtree. In particular, given a `Node`, I'd like to know its children.

I've seen that in this [SO question](https://stackoverflow.com/questions/37337758/does-boost-r-tree-support-hierarchy-traverse) there's al ink to the following line:

https://github.com/boostorg/geometry/blob/f92671b933e585ecba3c1b02c830f6e90e813da9/include/boost/geometry/index/detail/rtree/utilities/print.hpp#L160

where `n` is an `InternalNode`. From my understanding, `it->first` is the bounding box, while `it->second` is the indexable. However, I don't see how one can query the children of a specific node. Is there any way to find them?

Any help is highly appreciated, as I am completely stuck so far.

Best,
Lukas

**Comment by awulkiew** [comment]
The elements of internal nodes are pairs of bounding box and pointer to child node. So `it->second` is a pointer which can be used for depth-first tree traversal like this:

https://github.com/boostorg/geometry/blob/f92671b933e585ecba3c1b02c830f6e90e813da9/include/boost/geometry/index/detail/rtree/utilities/print.hpp#L178

Indexables (points, boxes or segments) can be extracted from elements of leaf nodes (aka values of the rtree) like this:

https://github.com/boostorg/geometry/blob/f92671b933e585ecba3c1b02c830f6e90e813da9/include/boost/geometry/index/detail/rtree/utilities/print.hpp#L194

**Reply by boblukas** [reply]
Thanks a lot, I have one last question: if I want to retrieve (not to just print) the value of leaf nodes, how can I do?

To give more context: the R-Tree I'm working with is constructed inserting `std::pair`s of Box and integers. My goal is to retrieve that pair, but I am not able to find a way to do this. In practice, I'd like to store what is printed on output after executing line 

https://github.com/boostorg/geometry/blob/f92671b933e585ecba3c1b02c830f6e90e813da9/include/boost/geometry/index/detail/rtree/utilities/print.hpp#L194

**Reply by awulkiew** [reply]
Elements of leafs are values of the Rtree, so your pair would be `*it`. `tr(*it)` would be a box, the same as `it->first`.
